web3.js/src/epoch-schedule.ts

```diff
@@ -40,6 +53,29 @@ export class EpochSchedule {
     this.firstNormalSlot = firstNormalSlot;
   }
 
+  getEpoch(slot: number): number {
+    return this.getEpochAndSlotIndex(slot)[0];
+  }
+
+  getEpochAndSlotIndex(slot: number): [number, number] {
+    if (slot < this.firstNormalSlot) {
+      const epoch =
+        trailingZeros(nextPowerOfTwo(slot + MINIMUM_SLOT_PER_EPOCH + 1)) -
+        trailingZeros(MINIMUM_SLOT_PER_EPOCH) -
+        1;
+
+      const epochLen = this.getSlotsInEpoch(epoch);
+      const slotIndex = slot - (epochLen - MINIMUM_SLOT_PER_EPOCH);
+      return [epoch, slotIndex];
+    } else {
+      const normalSlotIndex = slot - this.firstNormalSlot;
+      const normalEpochIndex = (normalSlotIndex / this.slotsPerEpoch) | 0;
+      const epoch = this.firstNormalEpoch + normalEpochIndex;
+      const slotIndex = normalSlotIndex % this.slotsPerEpoch | 0;
```

Comment:
nit: no truncation needed here

Reply:
There is one in rust:
https://github.com/solana-labs/solana/blob/18ec6756e49507c3945ab3d13d34dacbe897843b/sdk/program/src/epoch_schedule.rs#L144-L146
For the edge case where we rem with 0...

Reply:
Yeah good point, either way is fine then. I'm fine with expecting `slotsPerEpoch` to never be 0